### PR TITLE
Cleaned up concurrency management between handling packets and reapin…

### PR DESCRIPTION
**Pull request by oxtoacart**
…g origins/connections

### CODE REVIEW

File: base.go [hunk 1]
```diff
@@ -213,7 +212,6 @@ func newOrigin(p *proxy, transportProtocolName string, addr addr, upstream event
 		stack:           s,
 		linkID:          linkID,
 		channelEndpoint: channelEndpoint,
-		clients:         make(map[tcpip.FullAddress]*baseConn),
```
Comment (oxtoacart) [inline base.go]: Moved to tcpOrigin

File: ipproxy.go [hunk 2]
```diff
@@ -119,15 +131,17 @@ type Proxy interface {
 type proxy struct {
 	acceptedPackets int64
 	rejectedPackets int64
+	numTcpOrigins   int64
```
Comment (oxtoacart) [inline ipproxy.go]: We now track counts of such things using sync.Atomic.

File: ipproxy.go [hunk 3]
```diff
@@ -119,15 +131,17 @@ type Proxy interface {
...
+	numTcpConns     int64
+	numUdpConns     int64
 
 	opts       *Opts
 	proto      tcpip.NetworkProtocolNumber
 	downstream io.ReadWriter
 
-	tcpOrigins   map[addr]*origin
-	tcpOriginsMx sync.Mutex
```
Comment (oxtoacart) [inline ipproxy.go]: Mutex no longer used because all access to tcpOrigins and udpOrigins maps happens on a single goroutine.

File: ipproxy.go [hunk 4]
```diff
@@ -119,15 +131,17 @@ type Proxy interface {
...
-	udpConns     map[fourtuple]*udpConn
-	udpConnsMx   sync.Mutex
+	pktIn      chan ipPacket
```
Comment (oxtoacart) [inline ipproxy.go]: This makes it possible to decouple the reading of downstream packets from the processing of packets and reaping (which now happens in a single goroutine).

File: ipproxy.go [hunk 5]
```diff
@@ -199,20 +210,44 @@ func (p *proxy) copyToUpstream(icmpStack *stack.Stack, icmpEndpoint *channel.End
...
+	defer p.closeNow()
+	defer icmpStack.Close()
+	defer p.closeTCP()
+	defer p.closeUDP()
+
+	reapTicker := time.NewTicker(1 * time.Second)
+	defer reapTicker.Stop()
+
+	for {
+		select {
```
Comment (oxtoacart) [inline ipproxy.go]: Heart of the change. We now handle inbound packets as well as reaping on a single goroutine, so tcpConns and udpConns maps don't need mutexes.

File: ipproxy.go (hunk 4 above)
Comment (hwh33) [reply ipproxy.go]: Great improvement

**Comment by hwh33** [comment]
Looks good! (ignoring the CircleCI build failure as that was an unrelated issue)

**Comment by oxtoacart** [comment]
Thanks @hwh33!

**Pull request merged**